crates/wasm-mutate/README.md

```diff
@@ -0,0 +1,141 @@
+  given the same inputs as the original Wasm module.
+* **determinism:** `wasm-mutate` is deterministic, i.e., given the same input
+  Wasm module and the same seed, it always produces the same mutated
+  output Wasm module.
+* **libfuzzer integration**: `wasm-mutate` integrates well with mutation-based fuzzers like libFuzzer. It
+  reuses the fuzzer's raw input strings. `wasm-mutate` works with the
+  `LLVMFuzzerCustomMutator` hook and the
+  `libfuzzer_sys::fuzz_mutator!` macro.
+  ### Example
+  ```rust
+  #![no_main]
+
+    use libfuzzer_sys::fuzz_target;
+    use wasmparser::WasmFeatures;
+
+    fuzz_target!(|bytes: &[u8]| {
+        let _ = env_logger::try_init();
+
+        let mut seed = 0;
+        let (wasm, _config) = match wasm_tools_fuzz::generate_valid_module(bytes, |config, u| {
+            seed = u.arbitrary()?;
+            Ok(())
+        }) {
+            Ok(m) => m,
+            Err(_) => return,
+        };
+        log::debug!("seed = {}", seed);
+        let mutated_wasm = wasm_mutate::WasmMutate::default()
+            .seed(seed)
+            .fuel(1000)
+            .preserve_semantics(true)
+            .run(&wasm);
+        let mutated_wasm = match mutated_wasm {
+            Ok(w) => {
+                w
+            }
+            Err(e) => match e {
+                wasm_mutate::Error::NoMutationsApplicable => return,
+                e => panic!("Unexpected mutation failure: {}", e),
+            },
+        };
+
+        if log::log_enabled!(log::Level::Debug) {
+            std::fs::write("mutated.wasm", &mutated_wasm).expect("should write `mutated.wasm` okay");
+            if let Ok(mutated_wat) = wasmprinter::print_bytes(&mutated_wasm) {
+                std::fs::write("mutated.wat", &mutated_wat).expect("should write `mutated.wat` okay");
+            }
+        }
+
+        let features = WasmFeatures::default();
+        let mut validator = wasmparser::Validator::new();
+        validator.wasm_features(features);
+
+        let validation_result = validator.validate_all(&mutated_wasm);
+        log::debug!("validation result = {:?}", validation_result);
+        assert!(
+            validation_result.is_ok(),
+            "`wasm-mutate` should always produce a valid Wasm file"
+        );
+    });
+
+  ```
+
+* **test case reduction (WIP):** `wasm-mutate` can have the ability to restrict
+  mutations to only those that shrink the size of the Wasm module. If it is used
+  in this mode, `wasm-mutate` essentially becomes a Wasm test-case reducer. We
+  are currently working to provide a prototype of this feature as a separate
+  binary. The following pseudo-Rust provdes the general picture of it as an
+  standard hill-climbing algorithm.
+
+  ```rust
+    let wasmmutate = wasm_mutate::WasmMutate::default()
+            .seed(seed)
+            .fuel(1000)
+            .preserve_semantics(true)
+            .reduce(true);
+
+    while MAX_ITERATIONS > 0 {
+        wasm = wasmmutate.run(&wasm);
```

Comment:
Presumably there should be some kind of interestingness test here, or else we could just immediately assign the empty Wasm module to `wasm` and call it a day.